Compile error by using generic fn as arg of method of generic type

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.3.1 267e26b.acbfe88
**OS:** macos, macOS, 12.6, 21G115

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
pub struct Vec<T> {
mut:
	data &T    [required]
	cap  usize [required]
	len  usize [required]
}

pub fn new<T>() Vec<T> {
	return Vec<T>{
		data: unsafe { nil }
		cap: 0
		len: 0
	}
}

pub fn (ar &Vec<T>) iter() Iter<T> {
	return Iter<T>{
		v: unsafe { ar }
	}
}

pub struct Iter<T> {
mut:
	v   &Vec<T> [required]
	pos usize
}

pub fn (mut iter Iter<T>) next() ?&T {
	if iter.pos >= iter.v.len {
		return none
	}
	defer {
		iter.pos++
	}
	return unsafe { &iter.v.data[iter.pos] }
}

fn main() {
	mut goods := new<int>()
	goods.call_generic_fn(fn (a &int) bool {
		return *a > 1
	})
}

fn (arr Vec<T>) call_generic_fn(cb fn (&T) bool) {
	for val in arr.iter() {
		println(cb(val))
	}
}
```

**What did you expect to see?**
build successfully and no outputs.
 
**What did you see instead?**
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_502/generic_fn_as_fn_arg3.9365426858930545915.tmp.c:12097:25: error: implicit declaration of function 'main__Iter_T_int_next' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                _option_int_ptr _t2 = main__Iter_T_int_next(&_t1);
                                      ^
/tmp/v_502/generic_fn_as_fn_arg3.9365426858930545915.tmp.c:12097:25: note: did you mean 'main__Iter_T_int_next_T_int'?
/tmp/v_502/generic_fn_as_fn_arg3.9365426858930545915.tmp.c:12073:17: note: 'main__Iter_T_int_next_T_int' declared here
_option_int_ptr main__Iter_T_int_next_T_int(main__Iter_T_int* iter) {
                ^
/tmp/v_502/generic_fn_as_fn_arg3.9365426858930545915.tmp.c:12097:19: error: initializing '_option_int_ptr' (aka 'struct _option_int_ptr') with an expression of incompatible type 'int'
                _option_int_ptr _t2 = main__Iter_T_int_next(&_t1);
                                ^     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning and 2 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/for.v
+++ b/vlib/v/gen/c/for.v
@@ -358,7 +358,7 @@ fn (mut g Gen) for_in_stmt(node_ ast.ForInStmt) {
 			g.writeln('for (size_t $node.key_var = 0;; ++$node.key_var) {')
 		}
 		t_var := g.new_tmp_var()
-		receiver_typ := next_fn.params[0].typ
+		receiver_typ := g.unwrap_generic(next_fn.params[0].typ)
 		receiver_styp := g.typ(receiver_typ)
 		mut fn_name := receiver_styp.replace_each(['*', '', '.', '__']) + '_next'
 		receiver_sym := g.table.sym(receiver_typ)

--- a/vlib/v/tests/generics_for_in_iterate_test.v
+++ b/vlib/v/tests/generics_for_in_iterate_test.v
@@ -1,0 +1,50 @@
+pub struct Vec<T> {
+mut:
+	data &T    [required]
+	cap  usize [required]
+	len  usize [required]
+}
+
+pub fn new<T>() Vec<T> {
+	return Vec<T>{
+		data: unsafe { nil }
+		cap: 0
+		len: 0
+	}
+}
+
+pub fn (ar &Vec<T>) iter() Iter<T> {
+	return Iter<T>{
+		v: unsafe { ar }
+	}
+}
+
+pub struct Iter<T> {
+mut:
+	v   &Vec<T> [required]
+	pos usize
+}
+
+pub fn (mut iter Iter<T>) next() ?&T {
+	if iter.pos >= iter.v.len {
+		return none
+	}
+	defer {
+		iter.pos++
+	}
+	return unsafe { &iter.v.data[iter.pos] }
+}
+
+fn test_generics_for_in_iterate() {
+	mut goods := new<int>()
+	goods.call_generic_fn(fn (a &int) bool {
+		return *a > 1
+	})
+	assert true
+}
+
+fn (arr Vec<T>) call_generic_fn(cb fn (&T) bool) {
+	for val in arr.iter() {
+		println(cb(val))
+	}
+}

